### PR TITLE
Remote: Fix a bug that outputs of actions tagged with no-remote are u…

### PR DESCRIPTION
**Pull request by coeuvre**
…ploaded to remote cache when remote execution is enabled.

Fixes https://github.com/bazelbuild/bazel/issues/14900.

Also fixes an issue that action result from just remotely executed action is not saved to disk cache. The root cause is the action result is inlined in the execution response hence not downloaded through remote cache, hence not saved to disk cache. This results in the second build misses the disk cache, but it can still hit the remote cache and fill the disk cache. The third build can hit disk cache.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/remote/common/RemoteActionExecutionContext.java [hunk 1]
```diff
@@ -13,37 +13,71 @@
...
+    UPLOAD_INPUTS,
+    EXECUTE_REMOTELY,
+    UPLOAD_OUTPUTS,
+    DOWNLOAD_OUTPUTS,
+    UPLOAD_BES_FILES,
   }
 
-  /** Returns the {@link Type} of the context. */
-  Type getType();
+  private final Spawn spawn;
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/remote/common/RemoteActionExecutionContext.java:35]: Mark as `@Nullable`?

File: src/main/java/com/google/devtools/build/lib/remote/common/RemoteActionExecutionContext.java [hunk 2]
```diff
@@ -13,37 +13,71 @@
...
+  private final RequestMetadata requestMetadata;
+  private final NetworkTime networkTime;
+
+  @Nullable private ExecuteResponse executeResponse;
+  private Step step;
+
+  public RemoteActionExecutionContext(
+      Spawn spawn, RequestMetadata requestMetadata, NetworkTime networkTime) {
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/remote/common/RemoteActionExecutionContext.java:43]: Same as above.

Comment (tjgq) [reply src/main/java/com/google/devtools/build/lib/remote/common/RemoteActionExecutionContext.java:43]: Also - should we make the ctor private to enforce construction through one of the `create()` methods?

File: src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java [hunk 3]
```diff
@@ -258,7 +259,7 @@ static Command buildCommand(
   public static class RemoteAction {
     private final Spawn spawn;
     private final SpawnExecutionContext spawnExecutionContext;
-    private final RemoteActionExecutionContext remoteActionExecutionContext;
+    private final RemoteActionExecutionContext context;
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:262]: Is it a good idea to rename this, given that there are two different context objects?

File: src/main/java/com/google/devtools/build/lib/remote/common/RemoteActionExecutionContext.java (hunk 2 above)
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/remote/common/RemoteActionExecutionContext.java]: Done.

File: src/main/java/com/google/devtools/build/lib/remote/common/RemoteActionExecutionContext.java (hunk 1 above)
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/remote/common/RemoteActionExecutionContext.java]: Done.

**Review by tjgq — APPROVED** [review]

**Pull request closed without merging**